src/canvasdrawer.js

```diff
@@ -0,0 +1,1151 @@
+        }
+
+        context.translate(position.x + size.x / 2, 0);
+        if (tile.flipped) {
+            context.scale(-1, 1);
+        }
+        context.drawImage(
+            rendered.canvas,
+            0,
+            0,
+            sourceWidth,
+            sourceHeight,
+            -size.x / 2,
+            position.y,
+            size.x,
+            size.y
+        );
+
+        context.restore();
+    }
+
+    /**
+     * @private
+     * @inner
+     * Get the context of the main or sketch canvas
+     * @param {Boolean} useSketch
+     * @returns
+     */
+    _getContext( useSketch ) {
+        var context = this.context;
+        if ( useSketch ) {
+            if (this.sketchCanvas === null) {
+                this.sketchCanvas = document.createElement( "canvas" );
+                var sketchCanvasSize = this._calculateSketchCanvasSize();
+                this.sketchCanvas.width = sketchCanvasSize.x;
+                this.sketchCanvas.height = sketchCanvasSize.y;
+                this.sketchContext = this.sketchCanvas.getContext( "2d" );
+
+                // If the viewport is not currently rotated, the sketchCanvas
+                // will have the same size as the main canvas. However, if
+                // the viewport get rotated later on, we will need to resize it.
+                if (this.viewport.getRotation() === 0) {
+                    var self = this;
+                    this.viewer.addHandler('rotate', function resizeSketchCanvas() {
+                        if (self.viewport.getRotation() === 0) {
+                            return;
+                        }
+                        self.viewer.removeHandler('rotate', resizeSketchCanvas);
+                        var sketchCanvasSize = self._calculateSketchCanvasSize();
+                        self.sketchCanvas.width = sketchCanvasSize.x;
+                        self.sketchCanvas.height = sketchCanvasSize.y;
+                    });
+                }
+                this._updateImageSmoothingEnabled(this.sketchContext);
+            }
+            context = this.sketchContext;
+        }
+        return context;
+    }
+
+    /**
+     * @private
+     * @inner
+     * Save the context of the main or sketch canvas
+     * @param {Boolean} useSketch
+     * @returns
+     */
+    _saveContext( useSketch ) {
+        this._getContext( useSketch ).save();
+    }
+
+    /**
+     * @private
+     * @inner
+     * Restore the context of the main or sketch canvas
+     * @param {Boolean} useSketch
+     * @returns
+     */
+    _restoreContext( useSketch ) {
+        this._getContext( useSketch ).restore();
+    }
+
+    // private
+    _setClip(rect, useSketch) {
+        var context = this._getContext( useSketch );
+        context.beginPath();
+        context.rect(rect.x, rect.y, rect.width, rect.height);
+        context.clip();
+    }
+
+    // private
+    _drawRectangle(rect, fillStyle, useSketch) {
```

Comment:
Do we need both this and `drawDebuggingRect`?

Reply:
I'm not sure. `drawRectangle` was in the existing code, so I left it. The signatures are not the same so it's not as easy as just getting rid of `_drawRectangle` by replacing it.

Reply:
Since `_drawRectangle` is private we can just get rid of it if we're not using it anymore. If we do need them both, that's fine too, but maybe we should update them somehow to make it clear what the difference is.

Reply:
Looks like `_drawRectangle` is used in this scenario to act as an opaque background for the tiled image during loading:

```
        if ( tiledImage.placeholderFillStyle && tiledImage._hasOpaqueTile === false ) {
            var placeholderRect = this.viewportToDrawerRectangle(tiledImage.getBounds(true));
             ...
            this._drawRectangle(placeholderRect, fillStyle, useSketch);
        }
```

Which is distinct from `drawDebugRect()`. I can add an internal comment to this effect.

Makes me think this functionality should be added to the `webgl` viewer too. I'll have to think about how to make that happen most efficiently with the current pipeline.

Reply:
Gotcha... So it's a case of filling a rectangle vs stroking a rectangle... I admit I didn't look close enough to notice. Glad I asked though, so we can get that functionality in the WebGL drawer!

Reply:
I looked into this a bit further. The public API for this is via the `placeholderFillStyle` option:

https://github.com/openseadragon/openseadragon/blob/6abcc76a93ee3369267908f0a80641dc171440ed/src/openseadragon.js#L211-L215

The existing API is deeply entwined with `canvas` operations so I think it doesn't make a ton of sense to recreate all of that in the `webgl` drawer. Perhaps this could become as a canvas-only option? Or, if we want to support it in the `webgl` drawer, limit it to a solid color?

Reply:
Sounds good. I guess this gives us a chance to figure out how we differentiate between features supported in one drawer vs. all of them. I figure we should be explicit about it somehow.